### PR TITLE
[ROCm][Kimi-K3] Add merged MoE front

### PR DESCRIPTION
**Pull request by jiacao-amd**
## Summary

Integrates the Kimi-K3 merged MoE front from the companion AITER PR
(ROCm/aiter#5321).

Kimi-K3's MoE front is three separate projections of the same `x[M, 7168]`:
shared gate/up `7168 -> 1536`, router logits `7168 -> 896`, and the routed
latent `7168 -> 3584`. For supported token counts this PR replaces them with a
single BF16 GEMM into an FP32 accumulator (`7168 -> 6016`) plus a fused
split/SiTU epilogue. Routed experts are untouched and still run on the existing
modular AITER MXFP4 backend.

There is no feature environment variable. The merged path is selected
automatically when the model and AITER expose the required contract and the
token count is one of the validated shapes; everything else falls back to the
native front.

## Supported shapes

```text
1..16, 32, 48, 64, 80, 96, 112, 128, 192, 512, 1024, 1536, 2048
```

`32..192` is the MTP decode bucket. A pure-decode step submits
`num_seqs * (1 + num_speculative_tokens)` tokens, so with the usual
`num_speculative_tokens=3` any server past four concurrent sequences asks for a
count above 16.

`_KIMI_K3_MERGED_FRONT_MAX_TOKENS` is unchanged at 2048, so the workspace
allocation and its `front[:num_tokens]` slicing are untouched. `M=4096` and
`M=8192` are excluded: they measured 0.960x and 0.882x.

## Kernel results

Single-GPU gfx950, complete front, CUDA Graph:

| M | Speedup vs native front |
|---|---:|
| 1..16 | 1.645x – 2.140x |
| 512 | 1.224x |
| 1024 | 1.185x |
| 1536 | 1.162x |
| 2048 | 1.129x |

All tested shapes pass numerical checks, including exact router top-16
selection. The `32..192` bucket is not separately microbenchmarked; it is
covered by the numerical tests against a `torch.mm` reference and end to end
below.

## End-to-end results

MI355X, TP8, DSpark MTP (`num_speculative_tokens=3`). Both arms of each
comparison ran back to back on one node with identical server arguments, KV
configuration and client invocation.

### 8K/1K fixed sequence length

ISL 8192, OSL 1024, prefix caching disabled, 610 requests per arm, zero
failures.

| Concurrency | Decode M | Output throughput | Median TPOT |
|---:|---:|---|---|
| 1 | 4 | 130.1 -> 135.9 tok/s, **+4.43%** | 7.15 -> 6.83 ms, **-4.53%** |
| 4 | 16 | 369.9 -> 380.5 tok/s, **+2.87%** | 9.25 -> 8.97 ms, **-2.98%** |
| 8 | 32 | 582.8 -> 606.5 tok/s, **+4.07%** | 11.99 -> 11.31 ms, **-5.66%** |
| 16 | 64 | 799.8 -> 814.3 tok/s, **+1.82%** | 18.00 -> 17.65 ms, **-1.98%** |
| 32 | 128 | 1030.2 -> 1044.7 tok/s, **+1.41%** | 28.95 -> 28.55 ms, **-1.37%** |

Median TTFT moves by at most 0.58% in any cell: prefill in this workload is
8130-8190 tokens per step and never matches a supported shape, so the gain is
decode-only. Repeating two of these rows independently bounds their
reproducibility at 0.13-0.80 pp.

### AgenticX trace replay, 3600 s per cell

| Metric | C1 | C4 |
|---|---|---|
| Total throughput | 10212 -> 10533 tok/s, **+3.14%** | 19211 -> 19617 tok/s, **+2.11%** |
| Output throughput | 84.77 -> 87.75 tok/s, **+3.52%** | 139.20 -> 146.61 tok/s, **+5.33%** |
| P90 TPOT | 8.92 -> 8.47 ms, **-5.03%** | 13.52 -> 13.36 ms, **-1.19%** |
| P90 interactivity | 112.1 -> 118.1 tok/s, **+5.29%** | 73.9 -> 74.8 tok/s, **+1.20%** |

Read the total-throughput row with care: it does not measure this change.
Input tokens are over 99% of all tokens in this trace and 97.6% of them hit the
prefix cache, so total throughput is close to (input tokens handed to this arm)
/ wall clock. Trace replay assigns sessions dynamically and the two arms did
not get an identical token mix -- the optimized arm received 3.04% more input
tokens at C1 and 2.81% more at C4. Dividing the measured total-throughput delta
by each arm's input-token and duration ratio leaves a residual of +0.00% at C1
and +0.02% at C4, so that row is token mix and nothing else.

Output throughput and P90 TPOT are the rows that carry the effect: both are
normalized per output token, so neither depends on how much input the arm
happened to receive. Note that they disagree in shape with the total row at C4,
where output throughput is +5.33% against a total of +2.11% -- the smaller
number there is the noisier one, not the more conservative one.

## Accuracy

GSM8K, full 1319-question test split, 5-shot, greedy. This is the check the
numerical tests structurally cannot make: they compare the fused kernel against
a `torch.mm` reference built from the same merged weight, so they are blind to
the merged front slicing its router logits out of a `7168 -> 6016` accumulator
where the native front runs a separate `7168 -> 896` GEMM. Both are
BF16 x BF16 -> FP32, but the wider shape picks a different solution with a
different accumulation order, and that rounding difference can flip a borderline
top-16 expert selection.

Both arms ran back to back on one node at `max_num_seqs=8` with
`num_speculative_tokens=3` and block rejection sampling, so a saturated decode
step submits 32 tokens and the merged front is the decode path for effectively
the whole run.

| Filter | Baseline | Merged |
|---|---:|---:|
| strict-match | 0.9659 | 0.9719 |
| flexible-extract | 0.9651 | 0.9712 |

Both arms answer the same questions, so the paired view is the informative one:
1295 of 1319 answers agree, 16 flip to correct and 8 flip to incorrect, an exact
two-sided sign test on the 24 discordant pairs giving p = 0.15. The merged front
is therefore indistinguishable from the native front on accuracy. The +0.6 pp is
run-to-run nondeterminism, not a claimed improvement.

## Validation

- `ruff check` and `ruff format` on the changed files: passed.
- Kimi-K3 runner tests: `59 passed`.
- Companion AITER gfx950 tests: `17 passed`.
- TP8 server startup and CUDA Graph capture: passed. This model carries no
  `@support_torch_compile` in tree, so there is no compile step to exercise and
  CUDA Graph capture runs through breakable CUDA graph.

Companion AITER PR: https://github.com/ROCm/aiter/pull/5321

### CODE REVIEW

**Comment by jiacao-amd** [comment]
### GSM8K accuracy A/B

Adding the accuracy evidence this PR was missing. The numerical tests compare
the fused kernel against a `torch.mm` reference built from the same merged
weight, which is self-consistent and therefore blind to the one failure mode
specific to this change: the merged front slices its router logits out of a
`7168 -> 6016` accumulator where the native front runs a separate `7168 -> 896`
GEMM. Both are BF16 x BF16 -> FP32, but the wider shape picks a different
solution with a different accumulation order, and that rounding difference can
flip a borderline top-16 expert selection. Only an end-to-end score can see it.

Full 1319-question GSM8K test split, 5-shot, greedy. Both arms ran back to back
on one node from the same image, at `max_num_seqs=8` with
`num_speculative_tokens=3` and block rejection sampling, so a saturated decode
step submits 32 tokens and the merged front is the decode path for effectively
the whole run.

| Filter | Baseline | Merged |
|---|---:|---:|
| strict-match | 0.9659 | 0.9719 |
| flexible-extract | 0.9651 | 0.9712 |

Both arms answer the same questions, so the paired view is the informative one
rather than the unpaired +/- 0.005 binomial band: 1295 of 1319 answers agree, 16
flip to correct and 8 flip to incorrect, and an exact two-sided sign test on the
24 discordant pairs gives p = 0.15.

The merged front is indistinguishable from the native front on accuracy. The
+0.6 pp is run-to-run nondeterminism, not a claimed improvement.
